### DISCUSSION TITLE
Automerge strategy autodiscovery order and GitHub commit signing/verifying

### DISCUSSION THREAD

**Discussion by pballandras** [renovatebot/renovate#32062] [Suggest an Idea]
### Tell us more.

In a nutshell, GitHub has a limitation that it won't sign commits when it merges a PR via the `rebase` strategy. We first discovered this limitation experimentally when we realized that PRs Renovate merged automatically didn't generate "verified" commits from PR merges. We then, after investigation, stumbled onto [this](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification#signature-verification-for-rebase-and-merge):

> When using the Rebase and Merge option on a pull request, it's important to note that the commits in the head branch are added to the base branch without commit signature verification.

The way this limitation makes its way into Renovate is that in the `initRepo` function in the GitHub platform, there's a block of code that checks how Renovate can automatically merge a PR based on what type of merge the repo supports. This function will eagerly choose the `rebase` strategy.

```javascript
export async function initRepo({...}) {
    // ...
    // GitHub allows administrators to block certain types of merge, so we need to check it
    if (repo.rebaseMergeAllowed) {
      config.mergeMethod = 'rebase';
    } else if (repo.squashMergeAllowed) {
      config.mergeMethod = 'squash';
    } else if (repo.mergeCommitAllowed) {
      config.mergeMethod = 'merge';
    } else {
      // This happens if we don't have Administrator read access, it is not a critical error
      logger.debug('Could not find allowed merge methods for repo');
    }
    // ...
}
```

There's nothing inherently wrong here, but it does mean that no PR in a default (no configuration) repository on GitHub merged automatically by Renovate will generate a verified commit on the head branch on GitHub.

This said, there's a workaround to have commits verified: disable rebase merge in the repo settings and now you'll get signatures/verified commits back. While this work, it's not the best (and that is only a personal opinion). 

My proposition (which I already proposed in PR [#32016](https://github.com/renovatebot/renovate/pull/32016) before talking here, sorry, I was too eager 😅) is to change the order of the `if` block to mitigate this limitation by GitHub. We'd go from:

1. Rebase
2. Squash
3. Merge

to:

1. Squash
2. Merge
3. Rebase

What do you think?

**Comment by rarkins** [comment]
I'm ok with treating squash as the preferred merge method, but:
- Let's have this change in a major release (v39)
- Make sure it's not mixed up with github platformAutomerge, which has a separate approach to signing commits

**Reply by pballandras** [reply]
> Let's have this change in a major release (v39)

Sounds reasonable to me! It _is_ possibly breaking for people.

> Make sure it's not mixed up with github platformAutomerge

Yes, I'll make sure it's not mixed up. I'll check out how Renovate does it.

**Reply by pballandras** [reply]
Update on the not mixing up with platformAutomerge: for GitHub, it affects it in the way that it will use the merge method in the end, but, in a way, it will also make sure that commits resulting from a platform automerge won't use `rebase` by default and have the opportunity to sign/verify commits.

I think, long term, the best would be to implement https://github.com/renovatebot/renovate/issues/16884 (Pass `automergeStrategy` via `platformOptions` to `createPr`) for GitHub since it will give back the control to the end user, but I think this is a good step in the right direction.

**Comment by rarkins** [comment]
@pballandras are you sure that squash merges are signed? I know you found docs saying rebase ones aren't, but I haven't found anything which says squash merges _are_. You wrote:

> This said, there's a workaround to have commits verified: disable rebase merge in the repo settings and now you'll get signatures/verified commits back. While this work, it's not the best (and that is only a personal opinion).

But is that grounded in any docs or direct testing?

**Reply by pballandras** [reply]
I tested using Renovate where I disabled all merge options except for one (either merge or squash) and then let Renovate automerge. Commit was verified when squash or merge commit were used. I could open up a public repo and redo my testing using official app since I did my tests using a self-hosted version of Renovate. If you'd like that, I could find a moment to squeeze in this in my day.

**Reply by pballandras** [reply]
I'd like to say, using Mend's app is very nice compared to self-hosted, but I a digress: here are the results!

## Methodology

I opened a repo, installed Renovate and then added dependencies. It was a shiny new repo, completely default except for a Ruleset that asks for PR reviews (which I exempted Renovate, worked like a charm).

Between 2 Renovate automerges, I would restrict more and more the merge methods allowed:

- PR 3 had rebase, squash and merge
- PR 4 had squash and merge
- PR 5 had merge

Then I let Renovate merge.

## Result

PR 3 in my test repo https://github.com/pballandras/test-renovate/commit/9b3f588fbfc31097f5d82c18058699fdad73d96b
```
DEBUG: Branch status green (branch="renovate/dotenv-16.x")
DEBUG: Automerging #3 with strategy auto (branch="renovate/dotenv-16.x")
DEBUG: mergePr(3, renovate/dotenv-16.x) (branch="renovate/dotenv-16.x")
DEBUG: mergePr (branch="renovate/dotenv-16.x")
{
  "options": {
    "body": {
      "merge_method": "rebase"
    }
  }
  "url": "repos/pballandras/test-renovate/pulls/3/merge"
}

DEBUG: PR merged (branch="renovate/dotenv-16.x")
{
  "automergeResult": {
    "sha": "9b3f588fbfc31097f5d82c18058699fdad73d96b",
    "merged": true,
    "message": "Pull Request successfully merged"
  }
  "pr": 3
}
```
Resulting commit, unsigned/verified.

---

PR 4 in my test repo https://github.com/pballandras/test-renovate/commit/87279a9f3a512ee1c14bd4a60db181ad1a0cc3c8
```
DEBUG: Automerging #4 with strategy auto (branch="renovate/actions-checkout-4.x")
DEBUG: mergePr(4, renovate/actions-checkout-4.x) (branch="renovate/actions-checkout-4.x")
DEBUG: mergePr (branch="renovate/actions-checkout-4.x")
{
  "options": {
    "body": {
      "merge_method": "squash"
    }
  }
  "url": "repos/pballandras/test-renovate/pulls/4/merge"
}

DEBUG: PR merged (branch="renovate/actions-checkout-4.x")
{
  "automergeResult": {
    "sha": "87279a9f3a512ee1c14bd4a60db181ad1a0cc3c8",
    "merged": true,
    "message": "Pull Request successfully merged"
  }
  "pr": 4
}
```
Commit is signed/verified

---

PR 5 (again) in my test repo https://github.com/pballandras/test-renovate/commit/290c0350b2429efa2f55af1e83e1b1e2a812f63a
```
DEBUG: Branch status green (branch="renovate/octokit-monorepo")
DEBUG: Automerging #5 with strategy auto (branch="renovate/octokit-monorepo")
DEBUG: mergePr(5, renovate/octokit-monorepo) (branch="renovate/octokit-monorepo")
DEBUG: mergePr (branch="renovate/octokit-monorepo")
{
  "options": {
    "body": {
      "merge_method": "merge"
    }
  }
  "url": "repos/pballandras/test-renovate/pulls/5/merge"
}

DEBUG: PR merged (branch="renovate/octokit-monorepo")
{
  "automergeResult": {
    "sha": "290c0350b2429efa2f55af1e83e1b1e2a812f63a",
    "merged": true,
    "message": "Pull Request successfully merged"
  }
  "pr": 5
}
```
Commit is signed/verified

**Reply by rarkins** [reply]
Great, thank you very much for verifying it! In this case let's definitely switch to Squash as preferred

**Comment by pballandras** [comment]
A [PR](https://github.com/renovatebot/renovate/pull/32016) for this was merged and will be released in v39, this can be closed. Thanks @rarkins!
